File: src/app/Pedido.php
<?php

namespace App;
use Exception;


enum StatusPedido: string
{
    case PENDENTE = 'PENDENTE';
    case PAGO = 'PAGO';
    case ENVIADO = 'ENVIADO';
}

class Pedido
{
    private int $id;
    private StatusPedido $status;
    
    /** @var ItemPedido[]*/
    private array $itens;

    public function __construct(int $id, array $itens)
    {
        $this->id = $id;
        $this->itens = $itens;
        $this->status = StatusPedido::PENDENTE;
    }

    public function getId(): int
    {
        return $this->id;
    }

    public function getStatus(): StatusPedido
    {
        return $this->status;
    }

    public function getItens(): array
    {
        return $this->itens;
    }

    public function getValorTotal(): float
    {
        $total = 0;
        foreach ($this->itens as $item) {
            $total += $item->calcularSubtotal();
        }
        return $total;
    }

    public function pagar(): void
    {
        if ($this->status !== StatusPedido::PENDENTE) {
            throw new Exception('Um pedido só pode ser pago se estiver PENDENTE.');
        }
        if(empty($this->itens)) {
            throw new Exception('Um pedido deve conter pelo menos um item para ser pago.');
        }
        $this->status = StatusPedido::PAGO;
    }

    public function enviar(): void
    {
        if ($this->status !== StatusPedido::PAGO) {
            throw new Exception('Um pedido só pode ser enviado se estiver PAGO.');
        }
        $this->status = StatusPedido::ENVIADO;
    }
}

File: 1-modelagem-codigo-python/app/pedido.py
from enum import Enum

from app.item_pedido import ItemPedido


class StatusPedido(Enum):
    PENDENTE = 'PENDENTE'
    PAGO = 'PAGO'
    ENVIADO = 'ENVIADO'


class Pedido:
    def __init__(self, id: int, itens: list[ItemPedido]):
        self.__id = id
        self.__itens = itens
        self.__status = StatusPedido.PENDENTE

    def get_id(self) -> int:
        return self.__id

    def get_status(self) -> StatusPedido:
        return self.__status

    def get_itens(self) -> list[ItemPedido]:
        return self.__itens

    def get_valor_total(self) -> float:
        total = 0.0
        for item in self.__itens:
            total += item.calcular_subtotal()
        return total

    def pagar(self) -> None:
        if self.__status != StatusPedido.PENDENTE:
            raise Exception('Um pedido só pode ser pago se estiver PENDENTE.')
        if not self.__itens:
            raise Exception('Um pedido deve conter pelo menos um item para ser pago.')
        self.__status = StatusPedido.PAGO

    def enviar(self) -> None:
        if self.__status != StatusPedido.PAGO:
            raise Exception('Um pedido só pode ser enviado se estiver PAGO.')
        self.__status = StatusPedido.ENVIADO